src/operation/bulk_write.rs

```diff
@@ -0,0 +1,338 @@
+        self.models.get(index).ok_or_else(|| {
+            ErrorKind::InvalidResponse {
+                message: format!("invalid operation index returned from bulkWrite: {}", index),
+            }
+            .into()
+        })
+    }
+
+    fn get_inserted_id(&self, index: usize) -> Result<Bson> {
+        match self.inserted_ids.get(&index) {
+            Some(inserted_id) => Ok(inserted_id.clone()),
+            None => Err(ErrorKind::InvalidResponse {
+                message: format!("invalid index returned for insert operation: {}", index),
+            }
+            .into()),
+        }
+    }
+}
+
+/// A helper struct for tracking namespace information.
+struct NamespaceInfo<'a> {
+    namespaces: Vec<RawDocumentBuf>,
+    /// Cache the namespaces and their indexes to avoid traversing the namespaces array each time a
+    /// namespace is looked up or added.
+    cache: HashMap<&'a Namespace, usize>,
+}
+
+impl<'a> NamespaceInfo<'a> {
+    fn new() -> Self {
+        Self {
+            namespaces: Vec::new(),
+            cache: HashMap::new(),
+        }
+    }
+
+    /// Gets the index for the given namespace in the nsInfo list, adding it to the list if it is
+    /// not already present.
+    fn get_index(&mut self, namespace: &'a Namespace) -> (usize, usize) {
+        match self.cache.get(namespace) {
+            Some(index) => (*index, 0),
+            None => {
+                let namespace_doc = rawdoc! { "ns": namespace.to_string() };
+                let length_added = namespace_doc.as_bytes().len();
+                self.namespaces.push(namespace_doc);
+                let next_index = self.cache.len();
+                self.cache.insert(namespace, next_index);
+                (next_index, length_added)
+            }
+        }
+    }
+}
+
+impl<'a> OperationWithDefaults for BulkWrite<'a> {
+    type O = BulkWriteResult;
+
+    type Command = RawDocumentBuf;
+
+    const NAME: &'static str = "bulkWrite";
+
+    fn build(&mut self, description: &StreamDescription) -> Result<Command<Self::Command>> {
+        let max_doc_size: usize = Checked::new(description.max_bson_object_size).try_into()?;
+        let max_message_size: usize =
+            Checked::new(description.max_message_size_bytes).try_into()?;
+        let max_operations: usize = Checked::new(description.max_write_batch_size).try_into()?;
+
+        let mut command_body = rawdoc! { Self::NAME: 1 };
+        let options = match self.options {
+            Some(options) => bson::to_raw_document_buf(options),
+            None => bson::to_raw_document_buf(&BulkWriteOptions::default()),
+        }?;
+        bson_util::extend_raw_document_buf(&mut command_body, options)?;
+
+        let max_document_sequences_size =
+            max_message_size - OP_MSG_OVERHEAD_BYTES - command_body.as_bytes().len();
+
+        let mut namespace_info = NamespaceInfo::new();
+        let mut ops = Vec::new();
+        let mut current_size = 0;
```

Comment:
Suggest using `Checked` to prevent possible unsigned overflow when adding to `current_size`:

```suggestion
        let mut current_size = Checked::new(0);
```